framer/InputLayer.coffee

```diff
@@ -0,0 +1,148 @@
+Utils = require "./Utils"
+
+{Layer} = require "./Layer"
+{TextLayer} = require "./TextLayer"
+{Events} = require "./Events"
+
+Events.EnterKey  = "EnterKey"
+Events.BackspaceKey  = "BackspaceKey"
+Events.InputValueChange = "InputValueChange"
+Events.InputFocus = "InputFocus"
+Events.InputBlur = "InputBlur"
+
+class exports.InputLayer extends TextLayer
+
+	constructor: (options={}) ->
+
+		_.defaults options,
+			backgroundColor: "#FFF"
+			width: 500
+			height: 100
+
+		super options
+
+		# Layer containing input element
+		@input = new Layer
+			backgroundColor: "transparent"
+			x: 30
+			width: @width - 60
+			height: @height
+			parent: @
+			name: "input"
+
+		if @multiLine
+			@_inputElement = document.createElement("textarea")
+			@input.y = 25
+		else
+			@_inputElement = document.createElement("input")
+
+		# The id serves to differentiate multiple input elements from one another.
+		# To allow styling the placeholder colors of seperate elements.
+		@_id = Utils.round(Utils.randomNumber(0, 100))
+		@_inputElement.className = "input" + @_id
+
+		@on "change:width", =>
+			@_inputElement.style.width = "#{@input.width}px}"
+
+		# Append element
+		@input._element.appendChild(@_inputElement)
+
+		# Match TextLayer defaults and type properties
+		@_setTextProperties()
+
+		# All inherited properties
+		textProperties =
+			{@fontFamily, @fontSize, @lineHeight, @fontWeight, @color, @backgroundColor, @width, @height}
+
+		for property, value of textProperties
+
+			@on "change:#{property}", (value) =>
+				@_setTextProperties()
+				@_setPlaceholder()
+
+		# Set default placeholder text
+		@_defaultText = @text
+		@_setPlaceholder()
+
+		# Override text property setting the html
+		@html = ""
+
+		# Check if in focus
+		@_isFocused = false
+
+		# Default focus interaction
+		@_inputElement.onfocus = (e) =>
+
+			@focusColor = "#000"
+
+			# Emit focus event
+			@emit(Events.InputFocus, event)
+
+			@_isFocused = true
+
+		# Emit blur event
+		@_inputElement.onblur = (e) =>
+			@emit(Events.InputBlur, event)
+			
+			@_isFocused = false
+
+		# To filter if value changed later
+		currentValue = undefined
+
+		# Store current value
+		@_inputElement.onkeydown = (e) =>
+			currentValue = @value
+
+		@_inputElement.onkeyup = (e) =>
+
+			if currentValue isnt @value
+				@emit("change:value", @value)
+				@emit(Events.InputValueChange, @value)
+
+			# If enter key is pressed
+			if e.which is 13
+				@emit(Events.EnterKey, event)
+
+			# If backspace key is pressed
+			if e.which is 8
+				@emit(Events.BackspaceKey, event)
+
+	_setPlaceholder: =>
+		@_inputElement.placeholder = @_defaultText
+
+		unless @_isFocused
+			@_inputElement.style.color = @color ? "#aaa"
```

Comment:
Shouldn't this default be specified in a variable somewhere?
Also this can be shortened to:
```coffee
@_inputElement.style.color = @color ? "#aaa"
```
Notice the space between @color and the questionmark